src/include/btree_cmp_inline.h

```diff
@@ -14,12 +14,186 @@
+    t64 = firsteq ? (uint64_t)_mm_extract_epi64(t, 1) : tfirst;
+
+#ifndef WORDS_BIGENDIAN
+    u64 = __wt_bswap64(u64);
+    t64 = __wt_bswap64(t64);
+#endif
+
+    return (u64 < t64 ? -1 : u64 > t64 ? 1 : lencmp);
+}
+#else
+/*
+ * __lex_compare_gt_16 --
+ *     Lexicographic comparison routine for data greater than 16 bytes. Returns: < 0 if user_item is
+ *     lexicographically < tree_item, = 0 if user_item is lexicographically = tree_item, > 0 if
+ *     user_item is lexicographically > tree_item. We use the names "user" and "tree" so it's clear
+ *     in the btree code which the application is looking at when we call its comparison function.
+ *     Some platforms like ARM offer dedicated instructions for reading 16 bytes at a time, allowing
+ *     for faster comparisons.
+ */
+static WT_INLINE int
+__lex_compare_gt_16(const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp)
+{
+    struct {
+        uint64_t a, b;
+    } tdata, udata;
+    uint64_t t64, u64;
+    const uint8_t *tendp, *treep, *uendp, *userp;
+    bool firsteq;
+
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+
+    /*
+     * Compare 16 bytes at a time until we find a difference or run out of 16 byte chunks to
+     * compare.
+     */
+    for (userp = ustartp, treep = tstartp; uendp - userp > WT_VECTOR_SIZE;
+         userp += WT_VECTOR_SIZE, treep += WT_VECTOR_SIZE) {
+        memcpy(&udata, userp, WT_VECTOR_SIZE);
+        memcpy(&tdata, treep, WT_VECTOR_SIZE);
+        if (udata.a != tdata.a || udata.b != tdata.b)
+            goto final128;
+    }
+
+    /*
+     * Rewind until there is exactly 16 bytes left. We know we started with at least 16, so we are
+     * still in bounds.
+     */
+    memcpy(&udata, uendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+    memcpy(&tdata, tendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+
+final128:
+    firsteq = udata.a == tdata.a;
+    u64 = firsteq ? udata.b : udata.a;
+    t64 = firsteq ? tdata.b : tdata.a;
+
+#ifndef WORDS_BIGENDIAN
+    u64 = __wt_bswap64(u64);
+    t64 = __wt_bswap64(t64);
+#endif
+
+    return (u64 < t64 ? -1 : u64 > t64 ? 1 : lencmp);
+}
+#endif
+
+/*
+ * __lex_compare_le_16 --
+ *     Lexicographic comparison routine for data less than or equal to 16 bytes. Returns: < 0 if
+ *     user_item is lexicographically < tree_item, = 0 if user_item is lexicographically =
+ *     tree_item, > 0 if user_item is lexicographically > tree_item. We use the names "user" and
+ *     "tree" so it's clear in the btree code which the application is looking at when we call its
+ *     comparison function. Some platforms like ARM offer dedicated instructions for reading 16
+ *     bytes at a time, allowing for faster comparisons.
+ */
+static WT_INLINE int
+__lex_compare_le_16(const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp)
+{
+    uint64_t ta, tb, ua, ub, u64, t64;
+    const uint8_t *tendp, *uendp;
+
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+    if (len >> 3) {
```

Comment:
@quchenhao can you update the comment below:
`implicitly less than 128bits` -> `implicitly less than or equal 128bits`.

For consistency I'd also consider changing the `len & sizeof(uint32_t)` conditions in later branches to use `len >> 2` etc.

Reply:
There is a reason for using `if (var & SINGLE_BIT)`. Arm64 and x86 both have instructions to efficiently branch on that: `tbz`/`tbnz` on arm64 and `test`+`jz`/`jnz` on x86. The x86 one can take any bit pattern (so could efficiently handle `if (len >> 2)` with `~0b11`) but the arm version only allows you to test a single bit. As modified, `if (len >> 2)` would require a shift and a `cbz` instruction instead.

Reply:
Thanks for the context. I tried a [toy program](https://godbolt.org/z/fYqvqhscr) with `>>` and `&`, and `>>` looked cleaner to me, but based on the above `&` looks better. CC @quchenhao

Reply:
FYI, in your test the compiler actually eliminated the branch and emitted branchless code. That isn't possible in this context. [Here](https://godbolt.org/z/oxq73bTj3) is an example that shows the difference.

Also note that the compiler is turning `len >> 3` back into `len >= 8` because that is more efficient. It is also more clear to humans, so if we weren't going to do the single-bit check, we should write it the more natural way.

Reply:
I implemented a balanced approach. For the >= 64 bits case it tests (len & (sizeof(uint64_t) | sizeof(uint32_t))), in other case, it tests a single bit.

Reply:
A) I think you meant to use 8 | 16, rather than 4 | 8.
B) there is no benefit to that. Arm can only use tbz to test a single bit. You might as well do len >= 8 for the first test. It is more obviously correct and just as efficient as testing 2 bits. (But keep the remaining ones as single bit tests)